broadinstitute/single_cell_portal_core
# Image Pipeline: read gz files, handle 1M cells, add CLI arguments (SCP-4641)

This enables Image Pipeline to read gzipped files produced by the draft expression scatter data pipeline.  

Files will ultimately be decompressed in transit, we presume.  Decompressing them on the client is mostly a way to pick up draft scatter data pipeline output, until that other pipeline writes out individual gzipped files (without tarballing) and sets appropriate GCS metadata on them to enable [decompressive transcoding](https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding).

This also works around an issue in #1605 with using scatter files for the 1.3M cell study.  Using scatter files failed for that study, but not the 49k DE pilot study.  This seems caused not by the scatter data pipeline, but rather with using trimmed URLs to conditionally fetch only expression arrays (and not also annotation, cell, and coordinate) data in Image Pipeline.  The workaround still uses local data, but more than necessary, and can be optimized as a later, separate task.

To test:
* Install dependencies, if you hadn't in previous PR: `cd image-pipeline; yarn install`
* Find the accession for your local "Human milk -- differential expression" study
  * If you don't have that DE pilot study, you can add it via Terminal like so: `echo "SyntheticStudyPopulator.populate('human_milk_de_49k_cell')" | bundle exec rails console -e development`
* Confirm "progressive_loading" feature flag is _disabled_ for the study
* Download [All_Cells_UMAP_data.tar.gz](https://console.cloud.google.com/storage/browser/_details/broad-singlecellportal-public/test/All_Cells_UMAP_data.tar.gz), untar it, change untarred directory name from "All_Cells_UMAP-3d16b946-86a7-4d50-bce8-b97e2ab02734" to "All_Cells_UMAP"
* Ensure Rails is running (`rails s`)
* Changing "SCP138" to your DE pilot study accession and the `--json-dir` value to the path of your "All_Cells_UMAP" directory above, run:
`export NODE_TLS_REJECT_UNAUTHORIZED=0; time node expression-scatter-plots.js --accession="SCP138" --environment="development" --cores="1" --cluster="All_Cells_UMAP" --json-dir="/Users/eweitz/Downloads/All_Cells_UMAP/" --debug; tput bel # localhost, human milk DE study`
  * If you immediately see `TypeError: fetch failed`, edit `expression-scatter-plots.js` to use `https://127.0.0.1:3000` instead of `https://localhost:3000`, then re-run above
* Confirm `ls -latr output/<your DE pilot study accession>/images/`
  * If above times out, then simply try running above `expression-scatter-plots.js` command again

This satisfies SCP-4641.
